client/commands/query/get.go

```diff
@@ -51,65 +52,85 @@ func Get(key []byte, prove bool) (data.Bytes, uint64, error) {
+	*iavl.KeyExistsProof, *iavl.KeyNotExistsProof, error) {
 
 	resp, err := node.ABCIQuery("/key", key, true)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
 	}
-	ph := int(resp.Height)
 
 	// make sure the proof is the proper height
 	if !resp.Code.IsOK() {
-		return nil, 0, nil, errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
+		return nil, 0, nil, nil, errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
 	}
-	// TODO: Handle null proofs
-	if len(resp.Key) == 0 || len(resp.Value) == 0 || len(resp.Proof) == 0 {
-		return nil, 0, nil, lc.ErrNoData()
+	if len(resp.Key) == 0 || len(resp.Proof) == 0 {
+		return nil, 0, nil, nil, lc.ErrNoData()
 	}
-	if ph != 0 && ph != int(resp.Height) {
-		return nil, 0, nil, lc.ErrHeightMismatch(ph, int(resp.Height))
+	if resp.Height == 0 {
+		return nil, 0, nil, nil, errors.New("Height returned is zero")
 	}
 
-	check, err := GetCertifiedCheckpoint(ph)
+	check, err := getCertifiedCheckpoint(int(resp.Height), node, cert)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
+	}
+
+	if len(resp.Value) > 0 {
+		// The key was found, construct a proof of existence.
+		proof := new(iavl.KeyExistsProof)
+		err = wire.ReadBinaryBytes(resp.Proof, &proof)
+		if err != nil {
+			return nil, 0, nil, nil, err
+		}
+
+		// Validate the proof against the certified header to ensure data integrity.
+		err = proof.Verify(resp.Key, resp.Value, check.Header.AppHash)
+		if err != nil {
+			return nil, 0, nil, nil, err
+		}
+
+		return data.Bytes(resp.Value), resp.Height, proof, nil, nil
 	}
 
-	proof := new(iavl.KeyExistsProof)
+	// The key wasn't found, construct a proof of non-existence.
+	proof := new(iavl.KeyNotExistsProof)
 	err = wire.ReadBinaryBytes(resp.Proof, &proof)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, nil, err
 	}
 
-	// validate the proof against the certified header to ensure data integrity
-	err = proof.Verify(resp.Key, resp.Value, check.Header.AppHash)
+	// Validate the proof against the certified header to ensure data integrity.
+	err = proof.Verify(resp.Key, check.Header.AppHash)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, proof, errors.Wrap(err, "Couldn't verify proof")
 	}
 
-	return data.Bytes(resp.Value), resp.Height, proof, nil
+	return nil, resp.Height, nil, proof, lc.ErrNoData()
 }
 
-// GetCertifiedCheckpoint gets the signed header for a given height
+// getCertifiedCheckpoint gets the signed header for a given height
 // and certifies it.  Returns error if unable to get a proven header.
-func GetCertifiedCheckpoint(h int) (empty lc.Checkpoint, err error) {
-	// here is the certifier, root of all trust
-	node := commands.GetNode()
-	cert, err := commands.GetCertifier()
-	if err != nil {
-		return
-	}
-
-	// get the checkpoint for this height
+func getCertifiedCheckpoint(h int, node client.Client,
```

Comment:
I'm not familiar enough with the new codebase but seems like there's a bunch of global dependence we could/should avoid in these packages. So I like this direction of requiring the `node` to be passed in. Wonder if we can/should do that for more functions and relieve this package of importing `commands` altogether? I'll need to do more holistic review, just thinking aloud